Push notification bouncer

umairwaheed opened the pull request.

umairwaheed commented:
(Still need to clean commit messages in some commits.)

timabbott commented:
Code looks reasonable!  The main thing this needs is tests :).

timabbott commented on zerver/decorator.py:
```diff
@@ -417,12 +418,12 @@ def _wrapped_func_arguments(request, *args, **kwargs):
                 request._email = "zulip-server:" + role
                 profile.rate_limits = ""
+            else:
+                request._email = profile.email
```
I'm not totally on board with this change, since it loses the typing check we got from `isinstance`.

umairwaheed replied:
Fixed.

timabbott commented:
I merged the first half of the commits.  One thing I'm not sure I understand is this change:

`send_android_push_notification: Pass device token class argument.`

Since RemotePushDeviceToken is a subclass of PushDeviceToken, I think much of this change is unnecessary; leaving PushDeviceToken.GCM hardcoded there should be fine.

The rest looks reasonable, except that I don't see a proper end-to-end test for the actual push notification bouncing process (without mocking anything other than replacing network calls with calling the function in question).  Aka an extension to `test_push_bouncer_api` that actually bounces a push notification.

umairwaheed commented:
I still need to add a lot more tests.

umairwaheed commented:
@timabbott, ready for review.

umairwaheed commented:
(Actually don't merge yet; I need to clean the test commits.)

timabbott commented:
I tried deploying this, and found one fatal issue:
```
diff --git a/zerver/lib/push_notifications.py b/zerver/lib/push_notifications.py
index 1960418..4ccdc5b 100644
--- a/zerver/lib/push_notifications.py
+++ b/zerver/lib/push_notifications.py
@@ -32,9 +32,10 @@ import time
 import ujson
 from functools import partial
 
-try:
+
+if settings.ZILENCER_ENABLED:
     from zilencer.models import RemotePushDeviceToken
-except ImportError:  # nocoverage  -- Not convenient to add test for this.
+else:  # nocoverage  -- Not convenient to add test for this.
     from mock import Mock
     RemotePushDeviceToken = Mock()  # type: ignore # https://github.com/JukkaL/mypy/issues/1188
 
```

timabbott commented:
I also merged a couple commits fixing some logging-related issues caused by this branch; this may cause some small merge conflicts.

umairwaheed commented:
@timabbott, cleaned the commits. Now this can be merged if there isn't any issue with the code.

timabbott commented:
Merged, with a few commits on top to clean up code and fix a typing issue with Python 3.5.

Huge thanks for finishing this @umairwaheed!  

One follow-up is that I'm surprised that the line with the TODO comment I added in 0b2388bda92aea04bca1d4ca4e9cf490a96c0479 doesn't have test coverage; can you investigate?  I would expect that to have been tested by something like `test_end_to_end`.

umairwaheed commented:
>One follow-up is that I'm surprised that the line with the TODO comment I added in 0b2388bda92aea04bca1d4ca4e9cf490a96c0479 doesn't have test coverage; can you investigate? I would expect that to have been tested by something like test_end_to_end.

Sure.

umairwaheed commented:
>One follow-up is that I'm surprised that the line with the TODO comment I added in 0b2388b doesn't have test coverage; can you investigate? I would expect that to have been tested by something like test_end_to_end.

Fixed in https://github.com/zulip/zulip/pull/4813.
